File: lint/src/main/java/com/example/lint/LogDetector.java
package com.example.lint;

import com.android.tools.lint.detector.api.Category;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Implementation;
import com.android.tools.lint.detector.api.Issue;
import com.android.tools.lint.detector.api.Scope;
import com.android.tools.lint.detector.api.Severity;

import java.util.EnumSet;

public class LogDetector extends Detector {

    public LogDetector(){

    }
    public static final Issue ISSUE = Issue.create(
            "LogUsage",
            "不要在 AndroidManifest.xml 文件里同时设置方向和透明主题",
            "Activity 同时设置方向和透明主题在 Android 8.0 手机会 Crash",
            Category.SECURITY,
            5,
            Severity.ERROR,
            new Implementation(LogDetector.class,Scope.JAVA_FILE_SCOPE));


}

File: lint/src/main/java/com/example/lint/LogDetector.kt
package com.example.lint

import com.android.tools.lint.detector.api.*
import com.intellij.psi.JavaElementVisitor
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiMethodCallExpression
import org.jetbrains.uast.UCallExpression
import java.util.*


class LogDetector : Detector(), Detector.UastScanner {
    override fun getApplicableMethodNames(): List<String>? {
        return listOf("v", "d", "i", "w", "e", "wtf")
    }

    override fun visitMethodCall(context: JavaContext, node: UCallExpression, method: PsiMethod) {
        super.visitMethodCall(context, node, method)
        println("LogDetector:visitMethodCall ${method.name} ${context.getLocation(node).file.absolutePath}")
        if (context.evaluator.isMemberInClass(method, "android.util.Log")) {
            context.report(
                ISSUE,
                node,
                context.getLocation(node),
                "请勿直接调用android.util.Log，应该使用统一工具类"
            )
        }
    }

    companion object {
        @JvmField
        val ISSUE: Issue = Issue.create(
            "LogUsage",
            "避免调用android.util.Log",
            "请勿直接调用android.util.Log，应该使用统一工具类",
            Category.SECURITY, 5, Severity.ERROR,
            Implementation(LogDetector::class.java, Scope.JAVA_FILE_SCOPE)
        )
    }
}